src/lashup_kv.erl

```diff
@@ -229,72 +231,116 @@ init_db(Nodes) ->
+  Records1 = convert_records(Records0),
+  F = fun() ->
+        lists:foreach(fun(R) -> mnesia:write(R) end, Records1),
+        NClock = #nclock{key = node(), lclock = length(Records1)},
+        mnesia:write(NClock)
+      end,
+  case mnesia:sync_transaction(F) of
+    {atomic, _} ->
+      ok = mnesia:sync_log();
+    {aborted, Reason} ->
+      lager:error("Failed to upgrade kv table because ~p", [Reason]),
+      aborted
+  end.
+
+convert_records(Records) ->
+  convert_records(Records, []).
+
+convert_records([], Acc) ->
+  lists:reverse(Acc);
+convert_records([Record0|Rest], Acc) ->
+  Record1 = convert_record(Record0, length(Acc)),
+  convert_records(Rest, [Record1|Acc]).
+
+convert_record(_R = #kv{key = Key, vclock = VClock, map = Map}, Counter) ->
+  #kv2{key = Key, vclock = VClock, map = Map, lclock = Counter}.
+
+-spec(mk_write_fun(Key :: key(), OldVClock :: riak_dt_vclock:vclock() | undefined,
+      Op :: riak_dt_map:map_op(), LClock :: lclock()) -> (fun())).
+mk_write_fun(Key, OldVClock, Op, LClock) ->
   Node = node(),
   fun() ->
-    NewKV =
-      case mnesia:read(kv, Key, write) of
+    {NewKV, NClock} =
+      case mnesia:read(kv2, Key, write) of
         [] ->
           VClock = riak_dt_vclock:increment(Node, riak_dt_vclock:fresh()),
           Counter = riak_dt_vclock:get_counter(Node, VClock),
           Dot = {Node, Counter},
           {ok, Map} = riak_dt_map:update(Op, Dot, riak_dt_map:new()),
-          #kv{key = Key, vclock = VClock, map = Map};
-        [_ExistingKV = #kv{vclock = VClock}] when OldVClock =/= undefined andalso VClock =/= OldVClock ->
+          {#kv2{key = Key, vclock = VClock, map = Map, lclock = LClock},
+           #nclock{key = Node, lclock = LClock}};
+        [_ExistingKV = #kv2{vclock = VClock}] when OldVClock =/= undefined andalso VClock =/= OldVClock ->
           mnesia:abort(concurrency_violation);
-        [ExistingKV = #kv{vclock = VClock, map = Map}] ->
+        [ExistingKV = #kv2{vclock = VClock, map = Map}] ->
           VClock1 = riak_dt_vclock:increment(Node, VClock),
           Counter = riak_dt_vclock:get_counter(Node, VClock1),
           Dot = {Node, Counter},
           {ok, Map1} = riak_dt_map:update(Op, Dot, Map),
-          ExistingKV#kv{vclock = VClock1, map = Map1}
+          {ExistingKV#kv2{vclock = VClock1, map = Map1, lclock = LClock},
+           #nclock{key = Node, lclock = LClock}}
       end,
     case check_map(NewKV) of
       {error, Error} ->
         mnesia:abort(Error);
       ok ->
-        mnesia:write(NewKV)
+        mnesia:write(NewKV),
+        mnesia:write(NClock)
     end,
     NewKV
   end.
 
 -spec handle_op(Key :: term(), Op :: riak_dt_map:map_op(), OldVClock :: riak_dt_vclock:vclock() | undefined,
     State :: state()) -> {Reply :: term(), State1 :: state()}.
-handle_op(Key, Op, OldVClock, State0) ->
-  Fun = mk_write_fun(Key, OldVClock, Op),
+handle_op(Key, Op, OldVClock, State) ->
+  LClock = increment(getclock(node())),
```

Comment:
Shouldn't this LClock be fetched inside of the transaction that happens in mk_write_fun? Why are we passing LClock into mk_write fun, when we could have mk_write_fun fetch it inside of the same transaction?